drivebrain_core_base/include/public/MsgLogger.hpp

```diff
@@ -0,0 +1,148 @@
+        {
+            // should be fine for the most part to not lock here, this is the only place we are writing these bools
+            _running = false;
+            _logging = false;
+            _stop_log_function();
+            std::cout << "safely exited" << std::endl;
+        }
+
+        void log_msg(MsgType msg)
+        {
+            // TODO maybe make this also more thread safe ... ?
+            if (_logging)
+            {
+                auto out_msg = static_cast<std::shared_ptr<google::protobuf::Message>>(msg);
+                _handle_output_messages(msg, _logger_msg_function);
+            }
+            _handle_output_messages(msg, _live_msg_output_func);
+        }
+
+        // will only open a new file for logging if we are not currently logging
+        void start_logging_to_new_file()
+        {
+            std::unique_lock lk(_mtx);
+            if (!_logging)
+            {
+                _logging = true;
+                _current_log_name = _generate_log_name(_log_file_extension);
+                _open_log_function(_current_log_name);
+            }
+        }
+        std::tuple<const std::string &, bool> get_logger_status()
+        {
+            std::unique_lock lk(_mtx);
+            return {_current_log_name, _logging};
+        }
+
+        void stop_logging_to_file()
+        {
+            std::cout << "attempting stop" << std::endl;
+            {
+                std::unique_lock lk(_mtx);
+                if (_logging)
+                {
+                    _logging = false;
+                    _stop_log_function();
+                }
+            }
+            std::cout << "stopped" << std::endl;
+        }
+
+    private:
+        std::string _generate_log_name(const std::string &extension)
+        {
+            // Get the current time
+            auto t = std::time(nullptr);
+            auto tm = *std::localtime(&t);
+
+            // Create a stringstream to format the date and time
+            std::stringstream ss;
+
+            ss << std::put_time(&tm, "%Y-%m-%d_%H-%M-%S")
+               << extension;
+
+            // Return the generated filename
+            return ss.str();
+        }
+
+        void _handle_output_messages(MsgType msg, std::function<void(MsgType)> output_function)
+        {
+            auto out_msg = static_cast<std::shared_ptr<google::protobuf::Message>>(msg);
+            if (out_msg->GetDescriptor()->name() == "MCUOutputData")
+            {
+                auto cast_msg = std::static_pointer_cast<hytech_msgs::MCUOutputData>(out_msg);
+                if (cast_msg->brake_percent() == 0)
+                {
+                    std::cout << "empty msg recvd at output level" << std::endl;
+                }
+            }
+            output_function(msg);
+        }
+
+    private:
+        bool _running = true;
+        bool _logging = false;
+        std::mutex _mtx;
+        ThreadSafeOutput _thread_safe_log, _thread_safe_live_output;
```

Comment:
I don't think this is doing anything right now? Might want to remove for clarity.